Remove apply_legacy_finder_option methods

kbrock opened the pull request:
We were using `merge_conditions` to merge multiple where clauses together.
This behavior was deprecated in rails 4.0, but we had a legacy module included to merge where clauses together. We have successfully removed this by calling `where()` multiple times.
Now is time to remove.

In rails 4.0, the `find(:all, options)` was deprecated, preferring `where()`, `limit()` and others. `apply_legacy_finder_options` was a stop gap.
We no longer need to pass a hashe of conditions around.
Now is time to remove and just use standard rails `where()` methods.

/cc @chrisarcand Can't let you have ALL the fun in `Rbac` land
/cc @Fryguy not sure if you or @gtanzillo  should own this
/cc @lucasponce this is 1 of 4 for fixing #10087

jrafanie commented:
@kbrock Can you reword the title so someone naive (me) to your change has an idea what/why you're changing?

kbrock commented:
@jrafanie sorry. updated title, description, and removed cops

Fryguy commented on app/models/miq_report_result.rb:
```diff
@@ -71,7 +73,7 @@ def html_rows(options = {})
     update_attribute(:last_accessed_on, Time.now.utc)
     purge_for_user
-    html_details.apply_legacy_finder_options(options.merge(:order => "id asc")).collect(&:data)
+    html_details.order("id asc").offset(options[:offset]).limit(options[:limit]).collect(&:data)
```
Doesn't this ignore any other options that were passed in?  Or do to the callers never do that?

kbrock replied:
Very good question.

We only ever pass in `:per_page`, and `:page`.
We actually pass in both every time but one, when we generate the pdf.

Fryguy commented on app/controllers/application_controller.rb:
```diff
@@ -2230,15 +2230,15 @@ def find_by_id_filtered(db, id)
+      result = db.find_tags_by_grouping(mfilters, :ns => "*")
     else
-      result = db.apply_legacy_finder_options(options)
+      result = db
```
Same question here...if the callers passed in extra conditions they would be lost here (Can I assume since you removed the signature that was never happening...did you verify with only tests, or did you click around or grep?)

kbrock replied:
I used grep for both.

lucasponce commented:
@kbrock LiveMetric implemented just find() method to fetch data not where().
So, if find() will be decrecated I need to refactor the LiveMetric mechanism to support where().
It is not trivial, as I commented our use case is not to translate an ActsAsArModel into a physical table with supported SQL. In our case a virtual query is translated in specific calls to the hawkular backend.
So in this context, having a find() with all parameters just helped.
Anyway, I'm going to open an specific issue in LiveMetric to study/refactor the find() in where() calls.

lucasponce commented:
I have one question, sorry if I use this conversation, but it is related.
In LiveMetric we implement the find() method to fetch from a backend, and generate our virtual raws.
With this new approach, which is the equivalent?

kbrock commented:
@lucasponce implement find. You'll get a hash coming through to you. `ActsAsArQuery` converts the chainable form into the form you expect

lucasponce commented:
@kbrock that is what we have in place today https://github.com/ManageIQ/manageiq/blob/master/app/models/live_metric.rb#L6
but I tested the PR and it asks for the implementation of where()

lucasponce commented:
@kbrock I have take this PR and put a minor commit needed for testing in my branch:
https://github.com/lucasponce/manageiq/tree/x-livemetrics-finder-methods
Then, I have the following error:
https://gist.github.com/lucasponce/63070fedcf67226274fbd42ded28171a
which includes the part I don't get.
Following the stacktrace, it seems that is expected that the class livemetrics should implement the where().
Perhaps here 
https://github.com/kbrock/manageiq/blob/sql_munging_no_mo/lib/rbac/filterer.rb#L194
it should be some logic to invoke LiveMetrics.all ?

I will continue investigating it, but just in case if it is useful for you.
P.S.: I can give you access a hawkular provider to help you to reproduce the error.

lucasponce commented on lib/rbac/filterer.rb:
```diff
@@ -438,11 +434,12 @@ def to_class(klass)
+      klass = to_class(klass)
+
+      # could just call all on everything, but that will display deprecation warnings
+      klass = klass.all if klass < ActiveRecord::Base || klass.kind_of?(ActsAsArModel)
```
I have tested this locally and it fails:
  it "#check_acts_as_ar_model" do
    klass = LiveMetric
    expect(klass.kind_of?(ActsAsArModel)).to be(true)
  end
It seems that klass is of type Class and kind_of is checking some of ancestor of that class instead LiveMetric.
[Class, Module, RequireNested, MoreCoreExtensions::ModuleNamespace, Vmdb::ClassLogging, ActiveSupport::Dependencies::ModuleConstMissing, Module::Concerning, NakayoshiFork, ActiveSupport::ToJsonWithActiveSupportEncoder, Object, AvailabilityMixin::ClassMethods, AvailabilityMixin, PP::ObjectMixin, Vmdb::GlobalMethods, MoreCoreExtensions::ObjectNamespace, GettextI18nRails::HtmlSafeTranslations, FastGettext::Translation, ActiveSupport::Dependencies::Loadable, JSON::Ext::Generator::GeneratorMethods::Object, ActiveSupport::Tryable, Kernel, BasicObject]

kbrock commented:
@gtanzillo 

- rebased and resolved conflicts
- I moved the hawkular specific changes to another pr (that was the `klass.all` change)
- added: now always removing the limits. we add it back a few lines later but I don't trust the belongs to filter munging

Let me know if you want me to extract the first 3 commits out of here into separate PR

Fryguy commented:
@kbrock darga yes/no (probably will match whatever you say in #10264
